### PR TITLE
fix: code api changes

### PR DESCRIPTION
**Pull request by jozsef-armin-hamos**
### Description

_Provide description of this PR and changes, if linked Jira ticket doesn't cover it in full._

### Checklist

- [ ] Tests added and all succeed
- [ ] Linted
- [ ] README.md updated, if user-facing
- [ ] License file updated, if new 3rd-party dependency is introduced

### Screenshots / GIFs

_Visuals that may help the reviewer. Please add screenshots for any UI change. GIFs are most welcome!_

### CODE REVIEW

**Review by florinMiro — APPROVED** [review]

File: application/config/config.go [hunk 1]
```diff
@@ -561,7 +568,13 @@ func getCodeApiUrlFromCustomEndpoint(endpoint string) (string, error) {
 	}
 
 	m := regexp.MustCompile(`^(ap[pi]\.)?`)
-	endpointUrl.Host = m.ReplaceAllString(endpointUrl.Host, "deeproxy.")
+
+	subdomain := "deeproxy."
+	if isFedramp {
+		subdomain = "api."
```
Comment (michelkaporin) [inline application/config/config.go:574]: That might not work if a user provides a custom endpoint as `https://api.snykgov.io` in their settings (updated via [here](https://github.com/snyk/snyk-ls/blob/adc52f1731a191eef1b398ff3cffe6c1d1973224/application/config/config.go#L376)).

File: application/config/config_test.go [hunk 2]
```diff
@@ -161,43 +161,62 @@ func Test_loadFile(t *testing.T) {
...
 
 func TestSnykCodeApi(t *testing.T) {
+	t.Setenv("DEEPROXY_API_URL", "")
 	t.Run("endpoint not provided", func(t *testing.T) {
 
-		codeApiEndpoint, _ := getCodeApiUrlFromCustomEndpoint("")
+		codeApiEndpoint, _ := getCodeApiUrlFromCustomEndpoint("", false)
 		assert.Equal(t, "https://deeproxy.snyk.io", codeApiEndpoint)
+
+		codeApiEndpoint, _ = getCodeApiUrlFromCustomEndpoint("", true)
```
Comment (michelkaporin) [inline application/config/config_test.go:170]: This and test cases below can be simplified using table-driven tests. Using an array of test inputs and expected outputs for testing is more readable and maintainable.

File: application/config/config_test.go [hunk 3]
```diff
@@ -161,43 +161,62 @@ func Test_loadFile(t *testing.T) {
 }
 
 func TestSnykCodeApi(t *testing.T) {
+	t.Setenv("DEEPROXY_API_URL", "")
```
Comment (michelkaporin) [inline application/config/config_test.go:164]: Is it really needed? If yes, can we reuse `deeproxyApiUrlKey` const?

File: application/config/config.go [hunk 4]
```diff
@@ -866,6 +884,28 @@ func (c *Config) IsFedramp() bool {
...
+		return c.SnykCodeApi(), nil
+	}
+
+	// api url is fedramp, but code api url hasn't been updated properly
+	if strings.Contains(c.SnykCodeApi(), "deeproxy.") {
+		snykCodeApiUrl, err := getCodeApiUrlFromCustomEndpoint(c.SnykApi(), c.IsFedramp())
+		if err != nil {
+			return "", err
+		}
+		c.SetSnykCodeApi(snykCodeApiUrl)
```
Comment (michelkaporin) [inline application/config/config.go:901]: Getter functions shouldn't be mutating. As a caller of `Get` I expect it's not doing any side-effect mutation.

File: application/config/config.go [hunk 5]
```diff
@@ -866,6 +884,28 @@ func (c *Config) IsFedramp() bool {
 	host := strings.ToLower(u.Hostname())
 
 	// fedramp instance should have the format https://*.snykgov.io
-	snykgovDomain := strings.HasSuffix(host, ".snykgov.io")
-	return snykgovDomain
+	return isFedrampHost(host)
+}
+
+func (c *Config) GetSnykCodeApi() (string, error) {
```
Comment (michelkaporin) [inline application/config/config.go:890]: We have `SnykCodeApi` getter already at line :338.

File: application/config/config.go [hunk 6]
```diff
@@ -866,6 +884,28 @@ func (c *Config) IsFedramp() bool {
...
+	if !c.IsFedramp() {
+		return c.SnykCodeApi(), nil
+	}
+
+	// api url is fedramp, but code api url hasn't been updated properly
```
Comment (michelkaporin) [inline application/config/config.go:895]: When would that be a case? We should expect it to be updated (or do an update) by the time getter is called.

File: application/config/config.go [hunk 7]
```diff
@@ -866,6 +884,28 @@ func (c *Config) IsFedramp() bool {
...
+			return "", err
+		}
+		c.SetSnykCodeApi(snykCodeApiUrl)
+	}
+
+	if c.Organization() == "" {
+		return "", errors.New("Organization must be present in a fedramp environment")
+	}
+
+	path := "/org/" + c.Organization() + "/code"
```
Comment (michelkaporin) [inline application/config/config.go:908]: Will that work for when both org slug and org id are provided?

Comment (PeterSchafer) [reply application/config/config.go:908]: `c.Organization()` will always return the Org ID, no matter what the user specified. So this is fine!

File: application/config/config.go (hunk 5 above)
Comment (jozsef-armin-hamos) [reply application/config/config.go:890]: We do, but that is used by the sast_local_engine, and there is no error handling there, so I didn't want to introduce one. I thought the local_engine can use the original SnykCodeApi getter, and the backend_service will use the new one, that can also return an error.

File: application/config/config.go [hunk 8]
```diff
@@ -54,14 +55,15 @@ import (
...
-	DefaultDeeproxyApiUrl = "https://deeproxy.snyk.io"
-	pathListSeparator     = string(os.PathListSeparator)
-	windows               = "windows"
+	deeproxyApiUrlKey        = "DEEPROXY_API_URL"
+	FormatHtml               = "html"
+	FormatMd                 = "md"
+	snykCodeTimeoutKey       = "SNYK_CODE_TIMEOUT" // timeout as duration (number + unit), e.g. 10m
+	DefaultSnykApiUrl        = "https://snyk.io/api"
+	DefaultDeeproxyApiUrl    = "https://deeproxy.snyk.io"
+	DefaultSnykCodeApiGovUrl = "https://api.snyk.io"
```
Comment (PeterSchafer) [inline application/config/config.go:64]: Question: This seems redundant to `DefaultSnykApiUrl`, do we really need this? Or would it suffice to change the existing variable?

File: application/config/config.go (hunk 7 above)
Comment (michelkaporin) [reply application/config/config.go:908]: @PeterSchafer Thanks!

Can we document that in Code please? 🙏

**Pull request closed without merging**